<img src="https://www.unad.edu.co/images/footer/logo-unad-acreditacion-min.png" width="780" height="140" align="right"/>

<p align="center">Curso:Machine Learning</p>

<p align="center"> Código Curso: 203008067 </p>

<p align="center"> Grupo: 2 </p>

 <p align="center"> Fase 2 - Componente práctico - Algoritmos supervisados</p>

 <p align="center">  Presentado por: Fernando Arturo Varilla Mendoza </p>

<p align="center"> Código: 203008067A_1705 </p>

 <p align="center"> UNIVERSIDAD NACIONAL ABIERTA Y A DISTANCIA - UNAD </p>



# Acción 1: Aplicación modelos lineales.

In [3]:
# Paso 1 Cargamos la base de datos

import pandas as pd
import os
import numpy as np

data=pd.read_csv("Anexo1-zomato.csv",engine='python',encoding="latin-1")



Restaurant ID           Restaurant Name  Country Code              City  \
0           6317637          Le Petit Souffle           162       Makati City   
1           6304287          Izakaya Kikufuji           162       Makati City   
2           6300002    Heat - Edsa Shangri-La           162  Mandaluyong City   
3           6318506                      Ooma           162  Mandaluyong City   
4           6314302               Sambo Kojin           162  Mandaluyong City   
...             ...                       ...           ...               ...   
9546        5915730              NamlÛ± Gurme           208         ÛÁstanbul   
9547        5908749             Ceviz AÛôacÛ±           208         ÛÁstanbul   
9548        5915807                     Huqqa           208         ÛÁstanbul   
9549        5916112              Aôôk Kahve           208         ÛÁstanbul   
9550        5927402  Walter's Coffee Roastery           208         ÛÁstanbul   

                                                Address  \
0     Third Floor, Century City Mall, Kalayaan Avenu...   
1     Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   
2     Edsa Shangri-La, 1 Garden Way, Ortigas, Mandal...   
3     Third Floor, Mega Fashion Hall, SM Megamall, O...   
4     Third Floor, Mega Atrium, SM Megamall, Ortigas...   
...                                                 ...   
9546  Kemankeô Karamustafa Paôa Mahallesi, RÛ±htÛ±...   
9547  Koôuyolu Mahallesi, Muhittin íìstí_ndaÛô Cadd...   
9548  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9549  Kuruí_eôme Mahallesi, Muallim Naci Caddesi, N...   
9550  CafeaÛôa Mahallesi, BademaltÛ± Sokak, No 21/B,...   

                                        Locality  \
0      Century City Mall, Poblacion, Makati City   
1     Little Tokyo, Legaspi Village, Makati City   
2     Edsa Shangri-La, Ortigas, Mandaluyong City   
3         SM Megamall, Ortigas, Mandaluyong City   
4         SM Megamall, Ortigas, Mandaluyong City   
...                                          ...   
9546                                    Karakí_y   
9547                                   Koôuyolu   
9548                                 Kuruí_eôme   
9549                                 Kuruí_eôme   
9550                                        Moda   

                                       Locality Verbose   Longitude  \
0     Century City Mall, Poblacion, Makati City, Mak...  121.027535   
1     Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101   
2     Edsa Shangri-La, Ortigas, Mandaluyong City, Ma...  121.056831   
3     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.056475   
4     SM Megamall, Ortigas, Mandaluyong City, Mandal...  121.057508   
...                                                 ...         ...   
9546                                Karakí_y, ÛÁstanbul   28.977392   
9547                               Koôuyolu, ÛÁstanbul   29.041297   
9548                             Kuruí_eôme, ÛÁstanbul   29.034640   
9549                             Kuruí_eôme, ÛÁstanbul   29.036019   
9550                                    Moda, ÛÁstanbul   29.026016   

       Latitude                          Cuisines  ...          Currency  \
0     14.565443        French, Japanese, Desserts  ...  Botswana Pula(P)   
1     14.553708                          Japanese  ...  Botswana Pula(P)   
2     14.581404  Seafood, Asian, Filipino, Indian  ...  Botswana Pula(P)   
3     14.585318                   Japanese, Sushi  ...  Botswana Pula(P)   
4     14.584450                  Japanese, Korean  ...  Botswana Pula(P)   
...         ...                               ...  ...               ...   
9546  41.022793                           Turkish  ...  Turkish Lira(TL)   
9547  41.009847   World Cuisine, Patisserie, Cafe  ...  Turkish Lira(TL)   
9548  41.055817            Italian, World Cuisine  ...  Turkish Lira(TL)   
9549  41.057979                   Restaurant Cafe  ...  Turkish Lira(TL)   
9550  40.9

In [4]:
# filtro

df1= data[(data["City"] == "New Delhi") & (data["Votes"] > 0)]

# Dividimos el conjunto de datos

n=len(df1)
prueba=int(n * 0.20) # 20% de datos para probar
entrenamiento=n-prueba # 80% de datos para entrenar el modelo 


train_data = df1[:entrenamiento].copy()  # datos de entrenamiento
prueba_data = df1[entrenamiento:].copy()  # datos de prueba


# convertimos la variable independiente categorica a numerica
train_data['Has Online delivery'] = train_data['Has Online delivery'].map({'Yes': 1, 'No': 0})
prueba_data['Has Online delivery'] = prueba_data['Has Online delivery'].map({'Yes': 1, 'No': 0})


# Definimos las variables predictoras para entrenar el modelo
x_train = train_data[['Average Cost for two', 'Votes', 'Has Online delivery']]
y_train = train_data['Aggregate rating']

# Definimos las variables predictoras para probar el modelo
x_test = prueba_data[['Average Cost for two', 'Votes', 'Has Online delivery']]
y_test = prueba_data['Aggregate rating']



Restaurant ID           Restaurant Name  Country Code       City  \
2560       18287358                Food Cloud             1  New Delhi   
2561       18216944                 Burger.in             1  New Delhi   
2562         313333           Days of the Raj             1  New Delhi   
2563       18384127            Dilli Ka Dhaba             1  New Delhi   
2564            582                 Govardhan             1  New Delhi   
...             ...                       ...           ...        ...   
6949           7905               Moni Bakers             1  New Delhi   
6951           7818        Om Aggarwal Sweets             1  New Delhi   
6954       18372279          Midnight Delight             1  New Delhi   
6955       18384138  Ashok & Ashok Meat Dhaba             1  New Delhi   
6956       18211312       Durga Sweets Corner             1  New Delhi   

                                                Address        Locality  \
2560                              Aaya Nagar, New Delhi      Aaya Nagar   
2561        84, Near Honda Showroom, Adchini, New Delhi         Adchini   
2562  81/3, 1st Floor, Qutub Residency, Adchini, New...         Adchini   
2563  66 A, Ground Floor, Sri Aurobindo Marg, Adchin...         Adchini   
2564  84, Adjacent Hero Motor Bike Showroom, Main Me...         Adchini   
...                                                 ...             ...   
6949  5-A/1, Main Market, Madanpur Khadar, Near Agga...    Sarita Vihar   
6951  2 A, Pocket M & N, Janta Flats, Sarita Vihar, ...    Sarita Vihar   
6954  Near Sarita Vihar Metro Station, Sarita Vihar,...    Sarita Vihar   
6955       Shop 2, Ring Road, Sarojini Nagar, New Delhi  Sarojini Nagar   
6956  139, Near Ring Road Market, Sarojini Nagar, Ne...  Sarojini Nagar   

               Locality Verbose  Longitude   Latitude  \
2560      Aaya Nagar, New Delhi   0.000000   0.000000   
2561         Adchini, New Delhi  77.196923  28.535382   
2562         Adchini, New Delhi  77.197475  28.535493   
2563         Adchini, New Delhi  77.198033  28.537547   
2564         Adchini, New Delhi  77.196924  28.535523   
...                         ...        ...        ...   
6949    Sarita Vihar, New Delhi  77.297215  28.532575   
6951    Sarita Vihar, New Delhi  77.298525  28.538451   
6954    Sarita Vihar, New Delhi  77.288432  28.529367   
6955  Sarojini Nagar, New Delhi  77.193632  28.569428   
6956  Sarojini Nagar, New Delhi  77.194031  28.569744   

                                 Cuisines  ...            Currency  \
2560                       Cuisine Varies  ...  Indian Rupees(Rs.)   
2561                            Fast Food  ...  Indian Rupees(Rs.)   
2562   North Indian, Seafood, Continental  ...  Indian Rupees(Rs.)   
2563           South Indian, North Indian  ...  Indian Rupees(Rs.)   
2564  South Indian, North Indian, Chinese  ...  Indian Rupees(Rs.)   
...                                   ...  ...                 ...   
6949                               Bakery  ...  Indian Rupees(Rs.)   
6951                               Mithai  ...  Indian Rupees(Rs.)   
6954                North Indian, Chinese  ...  Indian Rupees(Rs.)   
6955                         North Indian  ...  Indian Rupees(Rs.)   
6956                  Mithai, Street Food  ...  Indian Rupees(Rs.)   

     Has Table booking Has Online delivery  Is delivering now  \
2560                No                   0                 No   
2561                No                   1                 No   
2562               Yes                   1                 No   
2563                No                   0                 No   
2564                No                   1                 No   
...                ...                 ...                ...   
6949                No                   0                 No   
6951                No                   0                 No   
6954                No                   1                 No   
6955                No                  

In [4]:
# cargar paquetes y aplicar el modelo lineal 

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

mod = LinearRegression()
mod.fit(x_train, y_train)

print("Coeficientes:", mod.coef_)
print("Intercepto:", mod.intercept_)

Coeficientes: [4.41814077e-04 5.53723965e-04 5.35120154e-01]
Intercepto: 2.329412489486186


In [6]:
# Hacer predicciones utilizando el conjunto de prueba
y_pred = mod.predict(x_test)
# Calcular el error cuadrático medio (MSE) y el coeficiente de determinación (R²)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("MSE:",mse)
print("R^2:",r2)



MSE: 1.690509529725134
R^2: 0.1545099023345221


# Acción 2: Aplicación modelo logístico.

In [2]:
# creacion de la nueva variable

df1 = data[(data["City"] == "New Delhi") & (data["Votes"] > 0)].copy()

# Crear la columna 'nivel_calificacion' sin la advertencia
df1['nivel_calificacion'] = df1['Aggregate rating'].apply(lambda x: 'alto' if x > 3.7 else 'bajo')



In [3]:
X = df1[['Average Cost for two', 'Votes', 'Has Online delivery']]
Y = df1['nivel_calificacion']

# Verificar el balance de la variable objetivo
print("Conteo de clases en 'nivel_calificacion':")
print(Y.value_counts())

Conteo de clases en 'nivel_calificacion':
bajo    4070
alto     688
Name: nivel_calificacion, dtype: int64


In [4]:
conteo_clases = Y.value_counts()
print("Conteo de clases en 'nivel_calificacion':")
print(conteo_clases)

# Si hay desbalance, hacer el balanceo con `iloc`
if conteo_clases['bajo'] != conteo_clases['alto']:  
    # Obtener el tamaño de la clase minoritaria
    menor_cantidad = conteo_clases.min()
    
    # Obtener índices aleatorios de ambas clases
    indices_alto = Y[Y == 'alto'].sample(n=menor_cantidad, random_state=42).index
    indices_bajo = Y[Y == 'bajo'].sample(n=menor_cantidad, random_state=42).index
    
    # Combinar los índices para obtener un conjunto balanceado
    indices_balanceados = np.concatenate([indices_alto, indices_bajo])
    
    # Seleccionar las filas balanceadas en X e y usando iloc
    X_balanced = X.loc[indices_balanceados]
    y_balanced = Y.loc[indices_balanceados]
else:
    X_balanced = X
    y_balanced = Y

# Verificar el balanceo final
print("Conteo de clases después del balanceo:")
print(y_balanced.value_counts())

Conteo de clases en 'nivel_calificacion':
bajo    4070
alto     688
Name: nivel_calificacion, dtype: int64
Conteo de clases después del balanceo:
alto    688
bajo    688
Name: nivel_calificacion, dtype: int64


In [5]:
n = len(df1)
prueba_size = int(n * 0.20)  # 20% de datos para prueba
entrenamiento_size = n - prueba_size  # 80% de datos para entrenamiento

# Crear conjuntos de datos
datos_entrenamiento = df1[:entrenamiento_size].copy()  # datos de entrenamiento
datos_prueba = df1[entrenamiento_size:].copy()  # datos de prueba

# Convertir 'Has Online delivery' a valores numéricos
datos_entrenamiento['Has Online delivery'] = datos_entrenamiento['Has Online delivery'].map({'Yes': 1, 'No': 0})
datos_prueba['Has Online delivery'] = datos_prueba['Has Online delivery'].map({'Yes': 1, 'No': 0})


# Separar las variables independientes (X) y la variable objetivo (y)
X_train = datos_entrenamiento[['Average Cost for two', 'Votes', 'Has Online delivery']]
Y_train = datos_entrenamiento['nivel_calificacion']
X_test = datos_prueba[['Average Cost for two', 'Votes', 'Has Online delivery']]
Y_test = datos_prueba['nivel_calificacion']


In [6]:
# ajuste del modelo logistico

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

mod2_logistico= LogisticRegression()
mod2_logistico.fit(X_train, Y_train)

# Predecir los valores en el conjunto de prueba
Y_pred = mod2_logistico.predict(X_test)

In [7]:
# coeficientes
coeficientes = mod2_logistico.coef_
intercepto = mod2_logistico.intercept_

# Imprimir los coeficientes y el intercepto
print("Coeficientes del modelo:", coeficientes)
print("Intercepto del modelo:", intercepto)

Coeficientes del modelo: [[-0.00080847 -0.00290371 -0.54591696]]
Intercepto del modelo: [3.11957037]


In [8]:
# Evaluación del modelo
precision = accuracy_score(Y_test, Y_pred)
# Imprimir los resultados
print("Precisión del modelo:", precision)


Precisión del modelo: 0.8759200841219769


# Acción 3: Aplicación K-NN.

In [20]:
# Teniendo en cuenta las variables definidas en el ejercisio anterior  entonces

from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV




KNeighborsClassifier(n_neighbors=16)

Dado que el modelo es basado en aprendisaje automatico, La estrategia a utilizar para seleccionar el mejor modelo se basa en en la búsqueda por cuadricula.Esta estrategia tiene como objetivo considerar diferentes valores del parametro  vecino o n_neighbors (k) el cual se refiere al número de puntos de datos más cercanos que el algoritmo considera al hacer una predicción. Esto se hace con el fin de  encontrar la mejor combinación de estos en términos de su precision. Para escoger los valores del rango me base en el referente  Applied Predictive Modeling y Machine Learning de Jason Brownlee el cual sugiere  un equilibrio entre el ajuste insuficiente (cuando k es demasiado bajo) y el sobreajuste (cuando k es demasiado alto).

In [24]:
modelo_knn = KNeighborsClassifier()
cuadricula= {'n_neighbors': range(1, 20)}  # Rango de valores de k 

# Configurar y ejecutar la búsqueda de hiperparámetros
grid_search = GridSearchCV(modelo_knn, cuadricula, cv=5)  # cv=5 realiza validación cruzada de 5 pliegues
grid_search.fit(X_train, Y_train)

# Mejor valor de k
mejor_k = grid_search.best_params_['n_neighbors']
print(f"El mejor valor de k es {mejor_k} con una precisión de {grid_search.best_score_:.4f}")

El mejor valor de k es 18 con una precisión de 0.8773


In [25]:
knn_model_final = KNeighborsClassifier(n_neighbors=18)  

# Ajustar el modelo
knn_model_final.fit(X_train, Y_train)

# Realizar predicciones en el conjunto de prueba

y_predictor = knn_model_final.predict(X_test)

# Calcular la precisión y mostrar la matriz de confusión
accuracy = accuracy_score(Y_test, y_predictor)
conf_matrix = confusion_matrix(Y_test, y_predictor)

print("Precisión del modelo KNN:", accuracy)
print("Matriz de confusión:\n", conf_matrix)

Precisión del modelo KNN: 0.8906414300736067
Matriz de confusión:
 [[ 45  87]
 [ 17 802]]


# Acción 4: Aplicación Árboles de decisión (CART).



In [10]:
# Teniendo en cuenta las variables definidas en el ejercisio anterior 
from sklearn.tree import DecisionTreeClassifier

modelo_arbol = DecisionTreeClassifier(random_state=42)
modelo_arbol.fit(X_train, Y_train)

# Hacer predicciones
predicciones = modelo_arbol.predict(X_test)

# Calcular la precisión
precision_arbol = accuracy_score(Y_test, predicciones)
print("Precisión del modelo de árbol de decisión:", precision_arbol)

Precisión del modelo de árbol de decisión: 0.8654048370136698


In [12]:
from sklearn.tree import export_graphviz
import graphviz

# Exportar el árbol
dot_data = export_graphviz(modelo_arbol, out_file=None, 
                           feature_names=['Average Cost for two', 'Votes', 'Has Online delivery'],  
                           class_names=['bajo', 'alto'],  
                           filled=True, rounded=True,  
                           special_characters=True)  
# Visualizar el árbol
grafico = graphviz.Source(dot_data)  
grafico.render("arbol_decision") 
grafico.view()  

ExecutableNotFound: failed to execute WindowsPath('dot'), make sure the Graphviz executables are on your systems' PATH

# Acción 5: Conclusiones.

**Modelo lineal**: De acuerdo a los coeficientes ajustado del modelo se puede decir que el interceptor se interpreta como una calificación promedio de 2.329 cuando no hay entrega en línea y los demás factores tienen valores nulos. Por otra parte, a medida que el costo promedio para dos personas aumenta, la calificación promedio solo incrementa en 0.00044 por cada unidad adicional. Del mismo modo, el número de votos tiene un efecto muy pequeño en la calificación, con un aumento de solo 0.00055 por cada voto adicional, sugiriendo que tampoco es un factor determinante. En cambio, la disponibilidad de entrega en línea tiene un impacto considerable en la calificación promedio, aumentando la calificación en 0.535 puntos cuando está disponible, es decir, la calificación promedio aumenta cuando la tienda tiene disponibilidad de entrega en línea, convirtiéndose en la variable más relevante.

Por ultimo Un MSE de 1.69 es relativamente regular, dado que los valores de la variable dependiente están entre un rango de 1 a 5, indicando que hay un gran margen de error en las predicciones del modelo.Esto se  corrobora con en R ajustado de 0.15 indicando que Solo un 15% de la variabilidad se explica por el modelo. En general el modelo no es significativamente bueno cuando se le ingresan datos nuevos por lo que ay que replantearlo, ingresar variables más relevantes o tomar un porcentaje de prueba diferente.

**Modelo logistico**: De acuerdo a los coeficientes del modelo se puede inferir que  la probabilidad de que una calificación se considere "alta" tiende a disminuir a medida que las variables aumenten, esto debido a que dichos coeficientes son negativos.Ademas, el coeficiente para Has Online delivery es significativo  lo que indica que los restaurantes que no ofrecen pedidos en línea tienen una probabilidad significativamente menor de tener calificaciones altas en comparación con aquellos que sí lo ofrecen.

Por otra parte al examinar el desempeño del modelo logistico, encontramos que este tiene un precision del 87.6% de clasificar correctamente todas las instancias, indicando que el modelo es eficaz. 

**Modelo KNN**:La precisión del modeloes del 89.06%, lo que indica que el modelo acierta en la clasificación de la mayoría de las instancias. Sin embargo, la matriz de confusión muestra que hubo 45 falsos negativos (casos de calificación baja clasificados como altos) y 87 falsos positivos (casos de calificación alta clasificados como bajos). A pesar de esto, el modelo tiene un  buen rendimiento en general, ya que se identificaron correctamente 802 casos de calificación alta.

**Arbol de decision**: El modelo de árbol de decisión clasifica correctamente el nivel de calificación como alto o bajo en el 86.54% de los casos en el conjunto de prueba. Esto sugiere que el modelo utiliza adecuadamente las variables de costo promedio, número de votos y si se ofrece entrega en línea para hacer predicciones. En el árbol de decisión visualizado, cada nodo representa una división basada en una de las variables: "Average Cost for two", "Votes", o "Has Online delivery". Los caminos de los nodos reflejan cómo el modelo toma decisiones para clasificar una instancia como "nivel_calificacion" alto o bajo. Los nodos finales, conocidos como hojas, contienen la clasificación final del modelo.

# Referencias

Brownlee, J. (2013). Machine Learning Mastery With Python: Understand Your Data, Create Accurate Models, and Work Projects End-to-End. Machine Learning Mastery.

VanderPlas, J. (2016). Python Data Science Handbook: Essential Tools for Working with Data. O'Reilly Media.

Raschka, S., & Mirjalili, V. (2017). Python Machine Learning - Second Edition: Vol. 2nd ed. Packt Publishing. https://bibliotecavirtual.unad.edu.co/login?url=https://search.ebscohost.com/login.aspx?direct=true&db=nlebk&AN=1606531&lang=es&site=eds-live&scope=site

Moran Pizarro, D. S. ., Dominguez Bonilla, S. J. ., Castaño Gutierrez, C. ., & Martinez Bez, C. E. . (2023). De Hilbert a los algoritmos cuánticos: el rol del álgebra en el desarrollo de la computación. Publicaciones E Investigación, 17(4). https://doi.org/10.22490/25394088.7503